### PR TITLE
Optimize query usage in the Onboarding tasks

### LINKED ISSUE #34833: [Enhancement]: Fix slow onboarding task list query
### Describe the solution you'd like

See https://github.com/woocommerce/woocommerce/issues/33988#issuecomment-1219915956

### Describe alternatives you've considered

_No response_

### Additional context

Still need to investigate.

### PR DESCRIPTION
**Pull request by moon0326**
### All Submissions:

-   [X] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [X] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR optimizes query usage in the Onboarding tasks. Thanks @sun  for the findings in https://github.com/woocommerce/woocommerce/issues/33988#issuecomment-1219915956

Closes #34833 

https://github.com/woocommerce/woocommerce/commit/78d05406679c4447c2319ae7801b7b778e24122a

`WC_Product_Query` generates the following query.

```
SELECT wp_posts.ID
FROM wp_posts
LEFT JOIN wp_term_relationships
ON (wp_posts.ID = wp_term_relationships.object_id)
WHERE 1=1
AND ( wp_term_relationships.term_taxonomy_id IN (2,3,4,5,16,17) )
AND wp_posts.post_type = 'product'
AND ((wp_posts.post_status = 'publish'))
GROUP BY wp_posts.ID
ORDER BY wp_posts.post_date DESC
LIMIT 0, 1
```

If you run `EXPLAIN` on the query, you will see that it uses `Using temporary; and Using filesort` due to GROUP BY and order by. This type of query gets slower as # of rows (in this case, products) grows.

`wp_count_posts` can be used instead. The query generated by `wp_count_posts` is simpler without any potential slowdown regardless of # of products.

```
SELECT post_status, COUNT( * ) AS num_posts
FROM wp_posts
WHERE post_type = 'product'
GROUP BY post_status
```

https://github.com/woocommerce/woocommerce/commit/3d0c0f7ca7f2633a77cf7c2c7e5d5cb438e652d0

Since we're trying to see if any of the tasks are incomplete, we can return early if we find an incomplete task.

https://github.com/woocommerce/woocommerce/commit/6e54edffb907ec3c15e4fd07bb0da947ef909789

`is_complete` is called multiple times in a single request. Since some tasks have DB calls in `is_complete`, it's a good idea to cache `is_complete`.

[Usage 1](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Admin/Features/OnboardingTasks/TaskLists.php#L434)

[Usage 2](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Admin/Features/OnboardingTasks/TaskList.php#L241)

### Other information:

-   [X] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [X] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> run changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Review by chihsuan — DISMISSED** [review]
Nice work! @moon0326 I've performed a few smoke tests on the task list and confirmed everything looks good. 👍

I observed that this PR reduced the # queries from 127 to 122. Though I wasn't able to see much difference in the local small site, believe it will improve the performance for large sites. 🚢  (Will approve PR again once we fixes E2E 😄 )

**Review by sun — DISMISSED** [review]

**Comment by moon0326** [comment]
> Nice work! @moon0326 I've performed a few smoke tests on the task list and confirmed everything looks good. 👍
> 
> I observed that this PR reduced the # queries from 127 to 122. Though I wasn't able to see much difference in the local small site, believe it will improve the performance for large sites. 🚢 (Will approve PR again once we fixes E2E 😄 )

Thank you! Could you approve one more time plz? CI has been fixed.

**Review by chihsuan — APPROVED** [review]
LGTM

File: plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/AdditionalPayments.php [hunk 1]
```diff
@@ -12,6 +12,13 @@
  */
 class AdditionalPayments extends Payments {
 
+	/**
+	 * Used to cache is_complete() method result.
+	 * @var null
+	 */
+	private $is_complete_result = null;
+
+
```
Comment (sun) [inline plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/AdditionalPayments.php:21]:
```suggestion
```

File: plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/Tax.php [hunk 2]
```diff
@@ -13,6 +13,12 @@
  */
 class Tax extends Task {
 
+	/**
+	 * Used to cache is_complete() method result.
+	 * @var null
+	 */
+	private $is_complete_result = null;
```
Comment (sun) [inline plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/Tax.php:20]: A static cache is a step forward. However, an actual cache entry set via `wp_cache_set()` would be preferable. We can invalidate the cache item(s) by implementing the action hook [`woocommerce_after_(object_type)_object_save`](https://wp-kama.com/plugin/woocommerce/hook/woocommerce_after_(object_type)_object_save) (for object_type "product").

**Review by sun — APPROVED** [review]

File: plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/Tax.php (hunk 2 above)
Comment (moon0326) [reply plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/Tax.php:20]: Thanks, @sun 

I think you meant to leave the comment for Products.php, not Tax.php.

With that said, I agree that caching the result is better. [wp_count_posts](https://github.com/WordPress/wordpress-develop/blob/6.0.2/src/wp-includes/post.php#L3051)  already caches the result using `wp_cache_set`.

**Pull request merged**